### ISSUE TITLE
DownloadButton getMimeType and dynamic file name

### ISSUE THREAD

**Issue by Joe32** [viritin/viritin#227]
Hi, I have two things:
1. https://github.com/viritin/viritin/blob/master/src/main/java/org/vaadin/viritin/button/DownloadButton.java#L96 idk maybe its not used anywhere but seems like it should be getMimeType
2. Is it possible to change filename of DownloadButton during creation of stream?

**Comment by mstahv** [comment]
I fixed the issue in getter.

Changing the file name on the fly is not easily achievable ATM, would need to get rid of using the whole Vaadin cores StreamSource stuff. Should handle download stuff completely internally, but that would on the otherhand make some things much simpler. I wont work on this now, unless many users find this relevant.

**Comment by Joe32** [comment]
Looks like it is working with UI#accessSynchronously. 
Another option could be overriding FileDownloader#handleConnectorRequest - this code sets http header: 
> `contentDisposition = "attachment; "
                        + DownloadStream.getContentDispositionFilename(stream
                                .getFileName());`

But its kinda hacky I'm ok with using accessSynchronously.

**Comment by Joe32** [comment]
Well both of my ideas does not work.

> Should handle download stuff completely internally, but that would on the otherhand make some things much simpler.

Can you elaborate, please? Could a custom servlet handle it?

**Comment by mstahv** [comment]
We should change this so that the download is handled by custom RequestHandler instead of StreamResource. Then on could set the name dynamically as well.

**Comment by mstahv** [comment]
This was fixed by @berndhopp and bit finetuned by me. It was actually easier than I thought. The solution is not perfect but somewhat works. The biggest gotcha is that you need to prepare for getFileName to be called earlier as well. It is also used for the URL, then later called again when using for content-disposition header, which actually defines the filename that browser uses.

See this usage example:
https://github.com/viritin/viritin/blob/master/src/test/java/org/vaadin/viritin/it/DownloadButtonExample.java

I'll try to cut a release with this fix in next 24h.

**Comment by Joe32** [comment]
Glat to hear that, had to quickfix with focusListener. Btw I also tried to implement RequestHandler in this button and result is here (Threads are deleted don't know how to deal with them neither are they needed):

    package org.vaadin.viritin.button;

    import com.vaadin.server.*;
    import com.vaadin.shared.MouseEventDetails;
    import org.vaadin.viritin.button.MButton;

    import java.io.IOException;
    import java.io.OutputStream;
    import java.math.BigInteger;
    import java.security.SecureRandom;


    /**
    * A helper class to implement typical file downloads.
    * <p>
    * With this class you'll get rid of lots of boilerplate code from your
    * application. It also inverts the bit cumbersome input-output stream API in
    * Vaadin so, normally you just "wire" this button to your backend method that
    * writes your resource to OutputStream (instead of playing around with piped streams or storing
    * resources temporary in memory. Example of usage:
    * <pre><code>
    *  new DownloadButton(invoice::toPdf).setFileName("invoice.pdf")
    * </code></pre>
    * <p>
    * The button extension hooks FileDownloader extension internally and inverts
    * the cumbersome default Vaadin API.
    * <p>
    * The writing of response is also spawn to separate thread, so in case your
    * resource generation takes lots of time, the UI wont block.
    */
    public class DownloadButton extends MButton {

        private final String resourceURL = "/dynamicResources/" + new BigInteger(130, new SecureRandom()).toString(32);
        private String mimeType;
        private long cacheTime = -1;
        private String fileName = "file";
        private int contentLength = -1;
        private ContentWriter writer;
        private final RequestHandler requestHandler = new RequestHandler() {
            @Override
            public boolean handleRequest(VaadinSession session,
                                        VaadinRequest request,
                                        VaadinResponse response)
                    throws IOException {
                if (resourceURL.equals(request.getPathInfo())) {
                    response.setContentType(mimeType);
                    response.setCacheTime(cacheTime);
                    if(contentLength > 0) {
                        response.setContentLength(contentLength);
                    }
                    response.setHeader("Content-Disposition", "attachment; filename=\"" + fileName + "\"");
                    writer.write(response.getOutputStream());
                    return true; // We wrote a response
                } else {
                    return false; // No response was written
                }
            }
        };
        /**
        * Constructs a new Download button without ContentWriter. Be sure to set
        * the ContentWriter or override its getter, before instance is actually
        * used.
        */
        public DownloadButton() {
        }

        public DownloadButton(ContentWriter writer) {
            this();
            this.writer = writer;
        }

        public ContentWriter getWriter() {
            return writer;
        }

        public DownloadButton setWriter(ContentWriter writer) {
            this.writer = writer;
            return this;
        }

        public String getMimeType() {
            return mimeType;
        }

        public DownloadButton setMimeType(String mimeType) {
            this.mimeType = mimeType;
            return this;
        }

        public String getFileName() {
            return fileName;
        }

        public DownloadButton setFileName(String fileName) {
            this.fileName = fileName;
            return this;
        }

        public long getCacheTime() {
            return cacheTime;
        }

        public DownloadButton setCacheTime(long cacheTime) {
            this.cacheTime = cacheTime;
            return this;
        }

        public int getContentLength() {
            return contentLength;
        }

        public void setContentLength(int contentLength) {
            this.contentLength = contentLength;
        }

        public String getResourceURL() {
            return resourceURL;
        }


        @Override
        public DownloadButton withIcon(Resource icon) {
            setIcon(icon);
            return this;
        }

        @Override
        public DownloadButton withClickShortcut    
    (int keycode, int... modifiers) {
            setClickShortcut(keycode, modifiers);
            return this;
        }

        @Override
        protected void fireClick(MouseEventDetails details) {
            super.fireClick(details);
            Page.getCurrent().open(VaadinServlet.getCurrent().getServletContext().getContextPath() + resourceURL, "");
        }

        @Override
        public void attach() {
            super.attach();
            VaadinSession.getCurrent().addRequestHandler(requestHandler);
        }

        @Override
        public void detach() {
            super.detach();
            VaadinSession.getCurrent().removeRequestHandler(requestHandler);
        }

        public interface ContentWriter {

            void write(OutputStream stream);
        }
    }
